cql3: convert LWT IF clause to expressions

avikivity opened the pull request:
LWT `IF` (column_condition) duplicates the expression prepare and evaluation code. Annoyingly,
LWT IF semantics are a little different than the rest of CQL: a NULL equals NULL, whereas usually
NULL = NULL evaluates to NULL.

This series converts `IF` prepare and evaluate to use the standard expression code. We employ
expression rewriting to adjust for the slightly different semantics.

In a few places, we adjust LWT semantics to harmonize them with the rest of CQL. These are pointed
out in their own separate patches so the changes don't get lost in the flood.

avikivity commented:
@cvybhu This is almost working (need to fix up the commits). Only problem (so far) is `IF some_column IN (1, 2, NULL)` since lists reject NULL.

I thought of several options, all look bad.

Here are the two least-worst:
1. create a new type list_with_nulls that does accept nulls (lots of work)
2. allow lists to accept nulls, only check when creating mutations out of them (risk: we'll miss a spot, and end up with NULLs in storage)
3. allow frozen lists to accept NULLs (in storage too), I think we can easily catch non-frozen lists during conversion to atomic cells

avikivity commented:
eh, that's three

avikivity commented:
I fixed up the commits, except the last two, which are still problematic.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3554/

nyh commented:
Issue #5825 about improper validation of nulls in list might be related.

avikivity commented:
> Issue #5825 about improper validation of nulls in list might be related.

Absolutely, I'll try to sanitize this area (before proceeding with the series).

avikivity commented:
> > Issue #5825 about improper validation of nulls in list might be related.
> 
> Absolutely, I'll try to sanitize this area (before proceeding with the series).

Turns out that's a driver bug.

avikivity commented:
Rebased on top of #12411 . Still some work to do (pass expression directly from parser).

/cc @gusev-p

avikivity commented:
Checkpoint: rebased past https://github.com/scylladb/scylladb/pull/12411

avikivity commented:
- rebased (thanks @cvybhu for binary_operator prepare_
- relaxed some LWT constraints/oddities in preparation
- tests now pass. 

The first six commits are part of https://github.com/scylladb/scylladb/pull/12411, so don't review them here.

avikivity commented:
Update:
 - drop empty patch (rebase artifact)
 - dropped comment about changed tests in next-to-last patch, those problems were identified and extracted to preceding patches

avikivity commented:
- rebased past merged 94331081585daffcf144cf2938d6d8bf1b42c150

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3935/

nyh commented:
Unfortunately I got confused and reviewed https://github.com/scylladb/scylladb/pull/12597 instead of this one :-( So after we finish the discussion there, I can merge this PR...

gusev-p commented:
> @cvybhu @gusev-p review ping

I'm looking into it.

cvybhu commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1329,6 +1329,10 @@ binary_operator prepare_binary_operator(binary_operator binop, data_dictionary::
     auto& prepared_lhs = *prepared_lhs_opt;
     lw_shared_ptr<column_specification> lhs_receiver = get_lhs_receiver(prepared_lhs, table_schema);
 
+    if (type_of(prepared_lhs)->references_duration() && is_slice(binop.op)) {
+        throw exceptions::invalid_request_exception(fmt::format("Duration type is unordered for {}", lhs_receiver->name));
+    }
+
```
I wonder if we also check types that contain duration, like `list<duration>`.
Maybe they fail at runtime?

avikivity replied:
references_duration() recurses into collection types and checks the element types.

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -212,7 +205,7 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
     } else {
-        throw exceptions::invalid_request_exception(format("subscripting non-map, non-list column {}", cdef->name_as_text()));
+        throw exceptions::invalid_request_exception(format("subscripting non-map, non-list column {}", s.val));
     }
```
Printing in error messages should use non-debug printing using `expression::printer`

avikivity replied:
I filed #12702 because I refuse to use expression::printer.

avikivity replied:
Fixed (using {:user}).

cvybhu commented on cql3/column_condition.cc:
```diff
@@ -123,82 +78,15 @@ bool column_condition::applies_to(const expr::evaluation_inputs& inputs) const {
-    }
-
-    auto& receiver = *def;
+    auto prepared = expr::prepare_expression(_expr, db, keyspace, &schema, make_lw_shared<column_specification>("", "", make_shared<column_identifier>("IF condition", true), boolean_type));
```
We could get the kespace and table name for `column_specification` constructor from the `schema`.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1483,6 +1544,10 @@ std::ostream& operator<<(std::ostream& s, oper_t op) {
         return s << "=";
     case oper_t::NEQ:
         return s << "!=";
+    case oper_t::EQ_SPECIAL_NULL:
+        return s << "EQUALS_SPECIAL_NULL";
+    case oper_t::NEQ_SPECIAL_NULL:
+        return s << "NOT_EQUALS_SPECIAL_NULL";
```
It's better to use `=` and `!=` to avoid surprising the users, even if we lost a little bit of information here.

nyh replied:
I agree. If the users used "=" in an expression, they would want to see "=" in the error message, not "EQUALS_SPECIAL_NULL" or "=*".
By the way, now that I think of it, I have some doubts if we really need the EQ_SPECIAL_NULL operator. Instead of being a different operator, it could have been an option to evaluate() on how to evaluate the regual EQ operator.

cvybhu replied:
I agree that a flag in `evaluation_inputs` that specifies in which context we are evaluating would be cleaner
I think the problem is that we would have to remember to set this flag in all the right places. With this solution we just generate the right data in the parser and then we don't have to worry that we might have missed setting the flag somewhere.
Another option would be having the flag inside `binary_operator`, something like `is_special_lwt_null_binop`

avikivity replied:
Changed to have a separate enum in binary_operator.

Printing now shows plain `=` in user mode, but annotated in debug mode.

avikivity commented on cql3/column_condition.cc:
```diff
@@ -102,170 +105,20 @@ column_condition::column_condition(const column_definition& column, std::optiona
-        return std::any_of(in_values.begin(), in_values.end(), [] (const bytes_opt& value) { return !value.has_value() || value->empty(); });
-    }
+    static auto true_value = raw_value::make_value(data_value(true).serialize());
+    return expr::evaluate(_expr, inputs) == true_value;
```
They're indeed not allowed.

gusev-p commented on cql3/expr/expression.cc:
```diff
@@ -247,41 +252,61 @@ bool_or_null equal(const expression& lhs, const managed_bytes_opt& rhs_bytes, co
+}
+
+static bool equal_special_null_impl(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
+   std::pair<managed_bytes_opt, managed_bytes_opt> sides_bytes =
```
I would prefer to have something like
```
enum class null_handling_style {
   standard,
   special
};
compare(managed_bytes_opt, managed_bytes_opt, null_handling_style) -> optional<bool_or_null>;
```
and not copy-paste `equal`, `not_equal`, `is_one_of_special_null`

avikivity replied:
I'm thinking we should move the null handling style into binary_operator, instead of adding new enums. It will make not changing </<=/>/>= meaning easier. Next revision will have it so.

gusev-p commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1242,6 +1243,84 @@ static lw_shared_ptr<column_specification> get_rhs_receiver(lw_shared_ptr<column
+                }
+            }
+        }
+        return subexpression;
```
By returning non-null we are not going inside the expr-tree, just check the top of it. Is it really what we want here?

avikivity replied:
search_and_replace recurses.

avikivity replied:
ah, but I broke the recursion. Will fix.

gusev-p commented on cql3/selection/selection.cc:
```diff
@@ -155,6 +155,18 @@ class simple_selection : public selection {
 
+shared_ptr<selection>
+selection_from_partition_slice(schema_ptr schema, const query::partition_slice& slice) {
+    std::vector<const column_definition*> cdefs;
```
`cdefs.reserve`?

avikivity replied:
Added.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -247,41 +252,61 @@ bool_or_null equal(const expression& lhs, const managed_bytes_opt& rhs_bytes, co
-static std::optional<std::pair<managed_bytes, managed_bytes>> evaluate_binop_sides(const expression& lhs,
+static std::pair<managed_bytes_opt, managed_bytes_opt> evaluate_binop_sides(const expression& lhs,
                                                                                    const expression& rhs,
                                                                                    const oper_t op,
```
It was unused before, so I'll leave it. Eventually we'll just drop evaluate_binop_sides, in fact it does very little already.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -435,30 +432,30 @@ bool_or_null like(const expression& lhs, const expression& rhs, const evaluation
+    expression lhs_constant = constant(raw_value::make_value(std::move(*lhs_bytes)), type_of(lhs));
+    utils::chunked_vector<managed_bytes_opt> list_elems = get_list_elements(raw_value::make_value(std::move(*rhs_bytes)));
     for (const managed_bytes_opt& elem : list_elems) {
         if (equal(lhs_constant, elem, evaluation_inputs{}).is_true()) {
```
Fixed outside the series.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -247,41 +252,61 @@ bool_or_null equal(const expression& lhs, const managed_bytes_opt& rhs_bytes, co
+    return !type_of(lhs)->equal(managed_bytes_view(*lhs_bytes), managed_bytes_view(*rhs_bytes));
+}
+
+static bool equal_special_null_impl(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
```
Removed.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2189,9 +2189,6 @@ static cql3::raw_value evaluate(const function_call& fun_call, const evaluation_
 
     for (const expression& arg : fun_call.args) {
         cql3::raw_value arg_val = evaluate(arg, inputs);
-        if (arg_val.is_null()) {
-            throw exceptions::invalid_request_exception(format("Invalid null value for argument to {}", *scalar_fun));
-        }
```
It's required indirectly, since we convert `x LIKE constant` to `func(x)`.

Functions handle NULL already, it's just this path that was broken (and isn't used in selection).

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -370,6 +374,15 @@ bool_or_null limits(const expression& lhs, oper_t op, const expression& rhs, con
             return false; // both NULL
         }
         break;
+    case oper_t::IN_SPECIAL_NULL:
+        if (!sides_bytes.second.has_value()) {
+            // Nothing can be IN a list if there is no list.
+            return false;
+        }
```
I changed `is_one_of()` (which now contains `is_one_of_special_null()`) to return `FALSE` for this case, to be more consistent with LWT, which doesn't return NULL (though it's rather invisible, since the condition is consumed internally and not exposed).

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -525,6 +538,27 @@ bool_or_null is_one_of(const expression& lhs, const expression& rhs, const evalu
+        return bool_or_null::null();
+    }
+    if (!sides_bytes.first) {
+        return boost::algorithm::any_of_equal(partially_deserialize_listlike(managed_bytes_view(*sides_bytes.second)), std::nullopt);
```
Fixed.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -525,6 +538,27 @@ bool_or_null is_one_of(const expression& lhs, const expression& rhs, const evalu
+    expression lhs_constant = constant(raw_value::make_value(std::move(*lhs_bytes)), type_of(lhs));
+    utils::chunked_vector<managed_bytes_opt> list_elems = get_list_elements(raw_value::make_value(std::move(*rhs_bytes)));
+    for (const managed_bytes_opt& elem : list_elems) {
+        if (equal(lhs_constant, elem, evaluation_inputs{}).is_true()) {
```
Fixed.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -1558,6 +1592,8 @@ std::ostream& operator<<(std::ostream& s, oper_t op) {
     case oper_t::IN:
         return s << "IN";
+    case oper_t::IN_SPECIAL_NULL:
+        return s << "IN_SPECIAL_NULL";
```
It's not `IN` in user mode and annotated in debug mode.

avikivity commented:
Update:


 - dropped `static` from function in anonymous namespace                                                                                                                                                                             
 - changed from SPECIAL_NULL_{EQ,NEQ,IN} to orthogonal enum                                                                                                                                                                          
 - restored < > <= >= LWT NULL semantics (throw if second operand is NULL)                                                                                                                                                           
   - fewer tests changed                                                                                                                                                                                                             
 - printout is now unchanged in user mode                                                                                                                                                                                            
 - is_one_of(x, NULL) now returns FALSE for LWT null handling (instead of                                                                                                                                                            
   NULL); doesn't affect anything since the result is consumed internally and                                                                                                                                                        
   not exposed to the user                                                                                                                                                                                                           
 - use list_contains_null() in one more place rather than open-coding it                                                                                                                                                             
 - call to evaluate(..., evaluation_inputs{}) (not forwarding real inputs)                                                                                                                                                           
   fixed (by eliminating the call site, is_one_of_special_null(), in favor of                                                                                                                                                        
   is_one_of())                                                                                                                                                                                                                      
 - fixed optimize_like() when LIKE is not a top-level expression; added unit                                                                                                                                                         
   test as penance                                                                                                                                                                                                                   
 - split removal of oper_is_slice() from restrictions.cc into its own patch,                                                                                                                                                         
   and justify the removal                                                                                                                                                                                                           
 - added vector::reserve() to selection_from_partition_slice()                                                                                                                                                                       
 - error message citing expression when subscripting non-map/non-list uses                                                                                                                                                           
   user-oriented expression printing                                                                                                                                                                                                 
 - disambiguated documentation of map element access with NULL subscript                                                                                                                                                             
 - adjusted cql-pytest to keep passing on Cassandra

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4289/

nyh commented on cql3/column_condition.cc:
```diff
@@ -72,15 +72,12 @@ int is_satisfied_by(cql3::expr::oper_t op, const abstract_type& cell_type,
-// Read the list index from key and check that list index is not
-// negative. The negative range check repeats Cassandra behaviour.
-uint32_t read_and_check_list_index(const cql3::raw_value_view& key) {
+// Read the list index from key. Negative values are allowed, like
```
I know this has already been decided, but I want to point out again that all these error cases like `list[-1]` being an error, were useful for catching buggy applications. Now `list[-1]` evaluates to null it will be harder for people who write applications to find bugs in the application.

avikivity replied:
It's part of the overall SQL philosophy. I propose we embrace it fully rather than find reasons to be different.

nyh commented on test/cql/lwt_test.result:
```diff
@@ -2045,7 +2045,11 @@ Error from server: code=2200 [Invalid query] message="Cannot store NULL in list
+| [applied]   | c       |
+|-------------+---------|
+| False       | [False] |
++-------------+---------+
```
Nitpick: to check if c[-1] is null, maybe it's better to check "c[-1] is null" instead of c[-1] = false?
(but I don't know, is IS NOT implemented in LWT? maybe it's not implemented).

avikivity replied:
IS NULL does't exist in LWT yet.

nyh commented on cql3/column_condition.cc:
```diff
@@ -34,64 +34,55 @@ void validate_operation_on_durations(const abstract_type& type, cql3::expr::oper
-            return false;
-        }
-}
+namespace cql3 {
```
Here too the mention of EQ_SPECIAL_NULL in the commit message is no longer correct.

avikivity replied:
Will fix.

nyh commented on cql3/column_condition.cc:
```diff
@@ -102,170 +93,20 @@ column_condition::column_condition(const column_definition& column, std::optiona
-    if (op != expr::oper_t::IN) {
-        assert(_in_values.empty());
-    }
+    _expr = update_for_lwt_null_equality_rules(_expr);
```
If I understand correctly, these three calls copy _expr three times, even if there is no need (e.g., optimize_like() will usually not do anything). Could these functions be changed to modify the expression in-place instead of always creating a new copy?

avikivity replied:
There's no way to modify in-place. It bothers me too, but of minor importance since it's done at prepare time.

nyh replied:
If there is no way to modify in-place, maybe there can at least be a way for a function that does nothing (e.g., if there is no LIKE at all) to at least not copy - maybe just move the old expression?

avikivity replied:
Yes, but why? We have to traverse the entire thing, so complexity is unchanged. It's not going to impact performance.

nyh replied:
Complexity doesn't change but may it slows this step by a large constant factor?
Anyway, I agree this can be optimized later.

avikivity replied:
These checks will make one case faster (when no changes need to be made) and another case faster (when we have to both check and rewrite). I think prepare should favor simplicity.

nyh reviewed (approved):
Looks good (and impressive).
There are two outdated commit messages refering to EQ_SPECIAL_NULL that you might want to fix.
I also left a bunch of minor nitpicks you may want to consider and possibly reject.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -259,9 +259,14 @@ static std::pair<managed_bytes_opt, managed_bytes_opt> evaluate_binop_sides(cons
 }
 
-bool_or_null equal(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
+bool_or_null equal(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs, null_handling_style null_handling) {
```
Will fix.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -302,16 +312,55 @@ bool limits(managed_bytes_view lhs, oper_t op, managed_bytes_view rhs, const abs
+                    throw exceptions::invalid_request_exception(fmt::format("Invalid comparison with null for operator \"{}\"", op));
+                } else {
+                    // LWT < > <= >= throws only for NULL second operand, not first
+                    return false;
```
Absolutely. The previous version eliminated the difference in favor of allowing NULL everywhere, but a reviewer objected and I restored the old behavior.

I plan to add `USING DIALECT OPTION = 'SQL' | 'LWT'` so people can move to SQL syntax and leave this quagmire beind.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1242,6 +1243,84 @@ static lw_shared_ptr<column_specification> get_rhs_receiver(lw_shared_ptr<column
 }
 
+class like_constant_function : public cql3::functions::scalar_function {
+    functions::function_name _name = { "system", "like(compiled)" };
```
I could change it to `like(<the original pattern>)`.

avikivity replied:
Changed.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -1242,6 +1243,84 @@ static lw_shared_ptr<column_specification> get_rhs_receiver(lw_shared_ptr<column
+
+    virtual sstring column_name(const std::vector<sstring>& column_names) const override {
+        return "LIKE";
+    }
```
It's part of function's interface. I propose not to change function's interface in this already long series.

nyh replied:
Ok.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2615,5 +2615,90 @@ unset_bind_variable_guard::is_unset(const query_options& qo) const {
+        }
+
+        virtual sstring column_name(const std::vector<sstring>& column_names) const override {
+            return "NONE";
```
A topic for another series.

nyh replied:
But these names "NONE" etc. are new to this series, aren't they? Or maybe I just missed it and it's pre-existing code that you moved?

avikivity replied:
These are overrides to an virtual functions defined by an abstract base class. I must define them.

avikivity commented on cql3/statements/modification_statement.cc:
```diff
@@ -169,13 +170,26 @@ bool modification_statement::applies_to(const update_parameters::prefetch_data::
-                value = &it->second;
+            val_opt = expr::extract_column_value(&cond->_column, inputs);
+        }
+        if (val_opt) {
```
I'll change it.

avikivity replied:
Changed. btw, this code was eliminated later, so the end result is unaffected by this.

avikivity commented:
Update:
 
 - reworded commit messages mentioning "SPECIAL NULL", which no longer exists
 - internal function for optimized like matcher now mentions the pattern it matches
 - rearranged modification_statement::applies_to()::condition_applies to
   eliminate an unnecessary check if the row happens to be missing

nyh commented:
Thanks. I'll wait for the CI to approve and merge this series, unless someone from the LWT team wants to wait so they can review this series as well? CC @kostja @gleb-cloudius

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4308/
